modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/SynonymsAnalysisTests.java

```diff
@@ -333,6 +335,109 @@ public void testChainedSynonymFilters() throws IOException {
+        Settings.Builder settingsBuilder = Settings.builder()
+            .put(IndexMetadata.SETTING_VERSION_CREATED, IndexVersion.current())
+            .put("path.home", createTempDir().toString());
+
+        String[] vocab = randomArray(50_000, 100_000, String[]::new, () -> randomAlphanumericOfLength(20));
+        int synonymsPerFilter = 10_000;
+        int synonymSets = 100;
+        List<String> filterNames = new ArrayList<>();
+        filterNames.add("lowercase");
+
+        for (int i = 1; i <= synonymSets; i++) {
+            String filterName = "synonyms_" + i;
+            StringBuilder sb = new StringBuilder();
+
+            for (int j = 0; j < synonymsPerFilter; j++) {
+                if (j > 0) {
+                    sb.append("\n");
+                }
+                for (int k = 0; k < between(1, 3); k++) {
+                    if (k > 0) {
+                        sb.append(", ");
+                    }
+                    for (int l = 0; l < between(1, 3); l++) {
+                        if (l > 0) {
+                            sb.append(" ");
+                        }
+                        sb.append(randomFrom(vocab));
+                    }
+                }
+
+                sb.append(" => ");
+                sb.append("syn").append(i * (j + 1)); // Shared ID appears in ALL filters
+            }
+
+            filterNames.add(filterName);
+            settingsBuilder.put("index.analysis.filter." + filterName + ".type", "synonym_graph")
+                .put("index.analysis.filter." + filterName + ".lenient", true)
+                .putList("index.analysis.filter." + filterName + ".synonyms", sb.toString());
+        }
+
+        settingsBuilder.put("index.analysis.analyzer.many_syn.tokenizer", "standard")
+            .putList("index.analysis.analyzer.many_syn.filter", filterNames);
+
+        Settings settings = settingsBuilder.build();
+        IndexSettings idxSettings = IndexSettingsModule.newIndexSettings("index", settings);
+
+        long startTime = System.currentTimeMillis();
+
+        // This would OOM without the SynonymGraphTokenFilterFactory::getSynonymFilter() fix (filters built sequentially)
+        indexAnalyzers = createTestAnalysis(idxSettings, settings, new CommonAnalysisPlugin()).indexAnalyzers;
+
+        // Verify the analyzer was built successfully and can analyze text
+        // With cross-referencing synonyms, the exact output is complex, so just verify it works
+        Analyzer analyzer = indexAnalyzers.get("many_syn");
+        assertNotNull("Analyzer should be created", analyzer);
+
+        for (int i = 0; i < 1000; i++) {
+            // Test that it can analyze without throwing exceptions
+            TokenStream ts = analyzer.tokenStream("test", randomFrom(vocab));
+            ts.reset();
+            assertTrue("Should produce at least one token", ts.incrementToken());
+            ts.close();
+        }
```

Comment:
Just to help the test GC along, you may want to place:

```
indexAnalyzers = null;
```

at the end of the test. (context: playing with the test, I had it loop through the entire test 20 times, and it OOM'd for me... adding this inside the loop, it passed). Not saying an OOM will happen during a CI test run with just the one time through in the test, but it can't hurt just in case.